plonky2-verifier/src/merkle/merkle_caps.cairo

```diff
@@ -0,0 +1,253 @@
+        loop {
+            i = 0;
+            // loop through the current level of internal nodes and compute hash(l, r) for each pair of nodes
+            while i < level_size {
+                let hash_val = hash_n_to_m_no_pad(
+                    array![*digests[start_idx + i], *digests[start_idx + i + 1]].span(), 1
+                )[0];
+                if (level_size / 2 == cap_size) {
+                    // cap is the next level - which means we no longer need to store 
+                    // the hashes of the level being computed in cap instead of digests
+                    cap.append(*hash_val);
+                } else {
+                    digests.append(*hash_val);
+                }
+                i += 2;
+            };
+
+            start_idx += level_size;
+            level_size /= 2;
+
+            if level_size == cap_size {
+                break;
+            }
+        };
+
+        MerkleTree { leaves: leaves, digests: digests, cap: MerkleCaps { data: cap } }
+    }
+
+    fn prove(self: @MerkleTree, index: usize) -> Array<Goldilocks> {
+        // includes the sibling of the leaf and the siblings of the nodes on the path to the cap level
+        let mut proof: Array<Goldilocks> = array![];
+        let mut i = index;
+
+        // add sibling of leaf to proof
+        if i % 2 == 0 {
+            proof.append(*self.leaves[i + 1]);
+        } else {
+            proof.append(*self.leaves[i - 1]);
+        };
+
+        // levels from the root to the cap level
+        let cap_height = self.cap.height();
+        // initialize to max number of levels in a normal merkle proof
+        let mut remaining_levels = log2_strict(self.leaves.len())
+            - 1; // -1 to account for the level of the leaf
+        let mut level_size = self.leaves.len() / 2;
+        let mut start_idx = 0; // index of the first node in the current level in the digests array
+
+        // add siblings of nodes on the path to the cap to proof
+        loop {
+            if remaining_levels == cap_height { // reached the cap level, no need to calculate up to the root
+                break;
+            }
+
+            i = i / 2; // parent index
+            if i % 2 == 0 { // left child
+                proof.append(*self.digests[start_idx + i + 1]);
+            } else {
+                proof.append(*self.digests[start_idx + i - 1]);
+            };
+
+            // move to the next level
+            start_idx += level_size;
+            level_size /= 2;
+            remaining_levels -= 1;
+        };
+
+        proof
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use plonky2_verifier::merkle::merkle_caps::MerkleCapsTrait;
+    use core::traits::Into;
+    use super::{gl, MerkleTreeImpl, MerkleCapsImpl};
+
+    #[test]
+    fn test_init() {
+        let leaves = array![gl(1), gl(2), gl(3), gl(4), gl(5), gl(6), gl(7), gl(8)];
+        let cap_size = 2;
+        let tree = MerkleTreeImpl::new(leaves, cap_size);
```

Comment:
Let's move all instances of this array and MerkleTree initialization into either a setup function or a const in mod tests to copy the tree variable from.